Adding missing features of cosine distribution

jaksle opened the pull request:
This is a very straightforward PR. It adds missing `mgf`, `cgf`, `cf` and direct affine transformations for the `Cosine` distribution. There are simple formulas for these, so there is no reason not to include them.

One thing which might require attention is if I protect from dividing by 0 correctly.

devmotion commented on src/univariate/continuous/cosine.jl:
```diff
@@ -93,3 +93,27 @@ function ccdf(d::Cosine{T}, x::Real) where T<:Real
+function mgf(d::Cosine{T}, t::Real) where T<:Real
+    μ, σ = params(d)
+    t ≈ 0. && return exp(μ*t)
+    return T(π^2) * exp(μ*t) * sinh(σ*t) / (σ*t*(T(π^2) + (σ*t)^2))
```
AFAICT the problem is not so much `t` in isolation but computing `sinh(σ*t) / (σ*t)` for small values of `σ*t`. Moreover, it seems a bit easier to cancel `π^2` and use the `IrrationalConstants.invπ` constant here. 

So I'd suggest something like

```suggestion
    σt = σ * t
    z = iszero(σt) ? one(float(σt)) : sinh(σ*t)/(σ*t)
    return exp(μ*t) * (z / (1 + (invπ * σt)^2))
```

jaksle commented on src/univariate/continuous/cosine.jl:
```diff
@@ -93,3 +93,27 @@ function ccdf(d::Cosine{T}, x::Real) where T<:Real
+function cgf(d::Cosine{T}, t::Real) where T<:Real
+    μ, σ = params(d)
+    t ≈ 0. && return μ*t
+    return 2log(π) + μ*t + σ*abs(t) + log1p(-exp(-2σ*abs(t))) - log(2) - log(σ*abs(t)) - log(T(π^2) + (σ*t)^2)
```
Great to know they have these useful functions there, thanks :)

jaksle commented:
I cleaned `cf` too. I also noticed now the parameters are used once, so their unpacking can be made even simpler.

I added testing against some known simple values. Some more are needed or is this enough?

My code was based how the rest of cosine.jl looked like. When I saw your corrections the code above started looking suspicious for me and indeed, `pdf`, `cdf`` and `ccdf` were type unstable. So I corrected them, though I'm not sure if my solution is optimal.

I'm also not sure about dependencies, but its' easy to change in any case.

jaksle commented:
@devmotion Could I ask for checking if now it is okay? I am a little unsure about the tests. Is there an instruction somewhere what is the recommended way to write them in Distributions.jl?

devmotion commented on src/univariate/continuous/cosine.jl:
```diff
@@ -59,37 +59,66 @@ kurtosis(d::Cosine{T}) where {T<:Real} = 6*(90-T(π)^4) / (5*(T(π)^2-6)^2)
+        z = (x - d.μ) / d.σ
+        return log1p(cospi(z)) - log(2d.σ)
+    else
+        return typemin(promote_type(T, S))
```
This is not correct, in general it creates a type instability.

devmotion commented on src/univariate/continuous/cosine.jl:
```diff
@@ -59,37 +59,66 @@ kurtosis(d::Cosine{T}) where {T<:Real} = 6*(90-T(π)^4) / (5*(T(π)^2-6)^2)
-        return zero(T)
-    end
+function ccdf(d::Cosine{T}, x::S) where {T<:Real, S<:Real}
+    W = promote_type(T, S)
```
Also here, that's not the correct return type.
